File: src/jkorn/pvpcore/PvPCore.php
<?php

declare(strict_types=1);

namespace jkorn\pvpcore;

/**
 * Created by PhpStorm.
 * User: jkorn2324
 * Date: 2019-04-04
 * Time: 16:01
 */

use jkorn\pvpcore\commands\CreateAreaCommand;
use jkorn\pvpcore\commands\EditWorldCommand;
use jkorn\pvpcore\commands\PvPAreaCommand;
use jkorn\pvpcore\commands\PvPCoreCommand;
use jkorn\pvpcore\utils\Utils;
use jkorn\pvpcore\world\areas\AreaHandler;
use pocketmine\plugin\PluginBase;
use jkorn\pvpcore\world\WorldHandler;


class PvPCore extends PluginBase
{

    // TODO FIX THESE BUGS:
    // - Loading all levels.
    // - Attack Speed not working.

    /** @var PvPCore */
    private static $instance;

    /** @var WorldHandler */
    private static $worldHandler;
    /** @var AreaHandler */
    private static $areaHandler;

    public function onEnable()
    {
        self::$instance = $this;

        $this->initCommands();

        // Loads all of the levels.
        Utils::loadLevels($this);

        self::$worldHandler = new WorldHandler($this);
        self::$areaHandler = new AreaHandler($this);

        new PvPCoreListener($this);
        $this->deleteConfig();
    }

    /**
     * Called when the PvPCore plugin is disabled.
     */
    public function onDisable()
    {
        if (self::$worldHandler instanceof WorldHandler) {
            self::$worldHandler->save();
        }

        if(self::$areaHandler instanceof AreaHandler) {
            self::$areaHandler->save();
        }
    }

    /**
     * @return PvPCore
     *
     * The static instance of the plugin.
     */
    public static function getInstance(): PvPCore
    {
        return self::$instance;
    }

    /**
     * @return WorldHandler
     *
     * The world handler.
     */
    public static function getWorldHandler(): WorldHandler
    {
        return self::$worldHandler;
    }

    /**
     * @return AreaHandler
     *
     * The area manager.
     */
    public static function getAreaHandler(): AreaHandler
    {
        return self::$areaHandler;
    }

    /**
     * Deletes the old config from the plugin folder
     * if it exists.
     */
    private function deleteConfig(): void
    {
        $configFile = $this->getDataFolder() . "config.yml";
        if (file_exists($configFile)) {
            unlink($configFile);
        }
    }

    /**
     * Initializes all of the commands.
     */
    private function initCommands()
    {
        Utils::registerCommand(new PvPCoreCommand());
        Utils::registerCommand(new EditWorldCommand());
        Utils::registerCommand(new CreateAreaCommand());
        Utils::registerCommand(new PvPAreaCommand());
    }
}

File: Nukkit/src/pvpcore/PvPCore.java
package pvpcore;

import cn.nukkit.event.Listener;
import cn.nukkit.plugin.PluginBase;
import pvpcore.commands.CreateAreaCommand;
import pvpcore.commands.EditWorldCommand;
import pvpcore.commands.PvPAreaCommand;
import pvpcore.commands.PvPCoreCommand;
import pvpcore.utils.Utils;
import pvpcore.worlds.WorldHandler;
import pvpcore.worlds.areas.AreaHandler;

import java.io.File;

public class PvPCore extends PluginBase implements Listener {

    /**
     * The worldHandler accessor.
     */
    private static WorldHandler worldHandler;
    /**
     * The areaHandler accessor.
     */
    private static AreaHandler areaHandler;

    /**
     * Called when the plugin is enabled.
     */
    @Override
    public void onEnable() {
        // Initializes the data folder.
        this.initDataFolder();
        // Initializes the Commands.
        this.initCommands();

        worldHandler = new WorldHandler(this);
        areaHandler = new AreaHandler(this);

        // Initializes the listener.
        new PvPCListener(this);

        this.deleteConfig();
    }

    /**
     * Called when the plugin is disabled.
     */
    @Override
    public void onDisable() {
        // Saves the worlds to the worlds file.
        if (worldHandler != null) {
            worldHandler.save();
        }

        // Saves the areas to the areas file.
        if (areaHandler != null) {
            areaHandler.save();
        }
    }

    /**
     * Gets the world handler.
     *
     * @return The world handler class.
     */
    public static WorldHandler getWorldHandler() {
        return worldHandler;
    }

    /**
     * Gets the area handler.
     *
     * @return The area handler class.
     */
    public static AreaHandler getAreaHandler() {
        return areaHandler;
    }

    /**
     * Initializes the data folder.
     */
    private void initDataFolder() {
        File file = this.getDataFolder();
        if (!file.exists() || !file.isDirectory()) {
            file.mkdir();
        }
    }


    /**
     * Deletes the config from the plugin data if it exists.
     */
    private void deleteConfig() {
        File configFile = new File(this.getDataFolder(), "config.yml");
        if (configFile.exists()) {
            configFile.delete();
        }
    }

    /**
     * Initializes the PvPCore commands.
     */
    private void initCommands() {
        Utils.registerCommand(new PvPCoreCommand());
        Utils.registerCommand(new PvPAreaCommand());
        Utils.registerCommand(new CreateAreaCommand());
        Utils.registerCommand(new EditWorldCommand());
    }
}
